src/scripts/upgrades/Upgrade.ts

```diff
@@ -0,0 +1,79 @@
+///<reference path="../GameConstants.ts"/>
+///<reference path="Cost.ts"/>
+///<reference path="CostFactory.ts"/>
+
+class Upgrade {
+    name: any;
+    displayName: string;
+    maxLevel: number;
+    _level: KnockoutObservable<number> = ko.observable();
+
+    // Describes whether this upgrade increases or decreases a number.
+    // (e.g. power is increasing, time is decreasing).
+    increasing: boolean;
+
+    // Optional array of costs
+    costList: Cost[] = [];
+    // Optional array of benefits
+    bonusList: number[] = [];
```

Comment:
Don't have to do this now. But we should think about if there's a need for upgrades with unlimited levels.

Reply:
This can be achieved by having the `bonusList` empty and overriding the `calculateBonus()` function to be infinite.

Reply:
Make sense.